common/cache/budget.go

```diff
@@ -0,0 +1,1260 @@
+	}
+
+	if m.logger != nil {
+		m.logger.Warn("Soft capacity limit exceeded",
+			tag.Name(m.name),
+			tag.Dynamic("cache_id", cacheID),
+			tag.Value(budgetType),
+			tag.Counter(int(requested)),
+			tag.Number(int64(available)),
+		)
+	}
+}
+
+// emitCapacityExceeded emits appropriate metrics based on the error type
+func (m *manager) emitCapacityExceeded(cacheID string, err error, requestedBytes uint64, requestedCount int64, capResult CapEnforcementResult) {
+	switch err {
+	case ErrBytesBudgetExceeded:
+		m.emitHardCapExceeded(cacheID, budgetTypeBytes, requestedBytes, capResult.AvailableBytes)
+	case ErrCountBudgetExceeded:
+		m.emitHardCapExceeded(cacheID, budgetTypeCount, uint64(requestedCount), uint64(capResult.AvailableCount))
+	case ErrBytesSoftCapExceeded:
+		m.emitSoftCapExceeded(cacheID, budgetTypeBytes, requestedBytes, capResult.AvailableBytes)
+	case ErrCountSoftCapExceeded:
+		m.emitSoftCapExceeded(cacheID, budgetTypeCount, uint64(requestedCount), uint64(capResult.AvailableCount))
+	}
+}
+
+// Cache-aware Reserve methods implementing two-tier soft cap logic
+
+func (m *manager) ReserveForCache(cacheID string, nBytes uint64, nCount int64) error {
+	// Check capacity constraints (both hard and soft cap)
+	capResult := m.enforceCapForCache(cacheID, nBytes, nCount)
+	if capResult.Error != nil {
+		m.emitCapacityExceeded(cacheID, capResult.Error, nBytes, nCount, capResult)
+		return capResult.Error
+	}
+
+	cacheUsage := m.getCacheUsage(cacheID)
+
+	cacheUsage.mu.Lock()
+	wasActive := cacheUsage.usedBytes > 0 || cacheUsage.usedCount > 0
+
+	if nBytes > 0 {
+		if err := m.reserveBytes(nBytes); err != nil {
+			cacheUsage.mu.Unlock()
+			return err
+		}
+		cacheUsage.usedBytes += nBytes
+		cacheUsage.fairShareCapacityBytes += capResult.FairShareBytes
+		cacheUsage.freeCapacityBytes += capResult.FreeBytes
+	}
+	if nCount > 0 {
+		if err := m.reserveCount(nCount); err != nil {
+			if nBytes > 0 {
+				m.releaseBytes(nBytes)
+				cacheUsage.usedBytes -= nBytes
+				cacheUsage.fairShareCapacityBytes -= capResult.FairShareBytes
+				cacheUsage.freeCapacityBytes -= capResult.FreeBytes
+			}
+			cacheUsage.mu.Unlock()
+			return err
+		}
+		cacheUsage.usedCount += nCount
+		cacheUsage.fairShareCapacityCount += capResult.FairShareCount
+		cacheUsage.freeCapacityCount += capResult.FreeCount
+	}
+
+	cacheUsage.mu.Unlock()
+
+	// Update active cache count after all operations succeed
+	m.updateActiveCacheCount(cacheUsage, wasActive)
+
+	// Update metrics after successful reservation
+	m.updateMetrics()
+
+	return nil
+}
+
+func (m *manager) ReserveBytesForCache(cacheID string, nBytes uint64) error {
+	// Check capacity constraints (both hard and soft cap)
+	capResult := m.enforceCapForCache(cacheID, nBytes, 0)
+	if capResult.Error != nil {
+		m.emitCapacityExceeded(cacheID, capResult.Error, nBytes, 0, capResult)
+		return capResult.Error
+	}
+
+	cacheUsage := m.getCacheUsage(cacheID)
+
+	cacheUsage.mu.Lock()
```

Comment:
same comment as  above: putting `defer unlock` next to the lock is generally safer / easier / works on panic recovery